matrix/Quaternion.hpp

```diff
@@ -325,6 +325,64 @@ class Quaternion : public Vector<Type, 4>
         return v * q  * Type(0.5);
     }
 
+    /**
+      * Computes the quaternion exponential of the 3D vector u
+      * as proposed in
+      * [1] Sveier A, Sjøberg AM, Egeland O. "Applied Runge–Kutta–Munthe-Kaas Integration
+      *     for the Quaternion Kinematics".Journal of Guidance, Control, and Dynamics. 2019 
+      *
+      * return a quaternion computed as
+      * expq(u)=[cos||u||, sinc||u||*u]
+      * sinc(x)=sin(x)/x in the sin cardinal function
+      *
+      * This can be used to update a quaternion from the body rates
+      * rather than using
+      * qk+1=qk+qk.derivative1(wb)*dt
+      * we can use
+      * qk+1=qk*expq(dt*wb/2)
+      * which is a more robust update.
+      * A re-normalization step might necessary with both methods.
+      *
+      * @param u 3D vector u
+      */
+	static Quaternion expq(const Vector3<Type> u)
+	{
+		const Type tol = Type(1.0e-6);
+		Type u_norm = u.norm();
+		Type sinc;
+		if (u_norm < tol) {
+			sinc = Type(1.0) - u_norm * u_norm / Type(6.0); // this ensure an error smaller than tol^3
+		} else {
+			sinc = Type(sin(u_norm) / u_norm);
```

Comment:
If this is supposed to run at high speed, should we maybe consider its Taylor series? 
![DeepinScreenshot_select-area_20210726173609](https://user-images.githubusercontent.com/14822839/127017245-2b497db9-7ad2-4696-9b28-8e5bdf3540ad.png)

Reply:
Isn't it what's implemented already?
in PX4-Autopilot/platforms/nuttx/NuttX/nuttx/libs/libc/math/[lib_sin.c](https://github.com/amlweems/wiced-history/blob/master/WICED/RTOS/NuttX/ver7.8/libc/math/lib_sin.c)

Reply:
FYI not that the distinction really matters here, but we use libm from the toolchain (newlib) and not what's directly in NuttX.

https://github.com/eblot/newlib/blob/master/newlib/libm/math/kf_sin.c

Reply:
> Isn't it what's implemented already?
in PX4-Autopilot/platforms/nuttx/NuttX/nuttx/libs/libc/math/lib_sin.c

Yes, but that's just for `sin(x)`; you then need to divide by `x` again. Using the Taylor series for `sin(x)/x` directly removes the need for any division (and you maybe don't even need to compute the 10 first elements?).

Reply:
ok, here is the idea:

if norm(u) is smaller than 0.2, we compute the 4 first terms of the Taylor series, which ensures an error smaller than 10^-10 for double (for float, the machine error is 10^-8 so the error is smaller than the machine error)
if norm(u) is bigger than 0.2, we call sin(u)/u and cos(u)

since the update is 
`qk+1=qk*expq(dt*wb/2)`
with dt about 4*10-3, wb can be as big as 100 rad/s in norm!

In practice such high body rates never happens, however I want to keep the correct formulation for higher norms just in case.

Reply:
Cool, I like it!